File: src/main/java/com/example/work/security/JwtTokenFilter.java
package com.example.work.security;

import com.example.work.exception.JwtAuthException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.GenericFilterBean;

import java.io.IOException;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@Component
@RequiredArgsConstructor
public class JwtTokenFilter extends GenericFilterBean {

    private final JwtTokenProvider jwtTokenProvider;

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        var token = jwtTokenProvider.resolveToken((HttpServletRequest) request);
        try {
            if (token != null && jwtTokenProvider.validateToken(token)) {
                var authentication = jwtTokenProvider.getAuthentication(token);
                if (authentication != null) {
                    SecurityContextHolder.getContext().setAuthentication(authentication);
                }
            }
        } catch (JwtAuthException e) {
            SecurityContextHolder.clearContext();
            ((HttpServletResponse) response).sendError(e.getHttpStatus().value());
            throw new JwtAuthException("JWT token is expired or invalid");
        }
        chain.doFilter(request, response);
    }
}

File: src/main/java/com/example/work/security/JwtTokenFilter.kt
package com.example.work.security

import com.example.work.exception.JwtAuthException
import lombok.RequiredArgsConstructor
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.GenericFilterBean
import javax.servlet.FilterChain
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


@Component
@RequiredArgsConstructor
class JwtTokenFilter(private val jwtTokenProvider: JwtTokenProvider) : GenericFilterBean() {

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        jwtTokenProvider.resolveToken(request as HttpServletRequest)?.let { token ->
            runCatching {
                jwtTokenProvider.getAuthentication(token)?.let {
                    SecurityContextHolder.getContext().authentication = it
                }
            }.onFailure {
                SecurityContextHolder.clearContext()
                sendError(response)
            }
        } ?: sendError(response)
        chain.doFilter(request, response)
    }

    private fun sendError(response: ServletResponse) {
        (response as HttpServletResponse).sendError(HttpStatus.FORBIDDEN.value())
        throw JwtAuthException("JWT token is expired or invalid")
    }
}
